PowerEditor/src/MISC/PluginsManager/Notepad_plus_msgs.h

```diff
@@ -549,6 +549,71 @@ enum Platform { PF_UNKNOWN, PF_X86, PF_X64, PF_IA64, PF_ARM64 };
 	// void* NPPM_GETBOOKMARKID(0, 0)
 	// Returns the bookmark ID
 
+	#define NPPM_DARKMODESUBCLASSANDTHEME (NPPMSG + 112)
+	// ULONG NPPM_DARKMODESUBCLASSANDTHEME(ULONG dmFlags, HWND hwnd)
+	// Subclass hwnd to add support for generic dark mode.
+	// When applying first time on main hwnd use "dmFlags = NppDarkMode::dmfAll".
+	// In case edit, listbox, static text, treeview, listview and toolbar controls don't have parent as main hwnd,
+	// call NPPM_DARKMODESUBCLASSANDTHEME on their parent and use "dmFlags = NppDarkMode::dmfSetParent".
+	// When handling dark mode change with NPPN_DARKMODECHANGED, use "dmFlags = NppDarkMode::dmfSetChildren".
+	// 
+	// Docking panels don't need to call NPPM_DARKMODESUBCLASSANDTHEME for main hwnd.
+	// Subclassing is applied automatically unless DWS_USEOWNDARKMODE flag is used.
+	//
+	// Might not work properly in C# plugins.
+	//
+	// Returns succesful combinations from dmfRequiredMask.
+	//
+
+	namespace NppDarkMode
+	{
+		// used on parent of edit, listbox, static text, treeview, listview and toolbar controls
+		constexpr ULONG dmfSetParent =          0x00000001UL;
+		// without dmfSubclassChildren will only theme
+		// should be handled on first time and in NPPN_DARKMODECHANGED
+		// only theme requires at least Windows 10
+		constexpr ULONG dmfSetChildren =        0x00000002UL;
+		// should be used only on main hwnd when applying subclass first time
+		constexpr ULONG dmfSubclassChildren =   0x00000004UL;
+		// set dark title bar
+		// should be handled on first time and in NPPN_DARKMODECHANGED
+		// requires at least Windows 10
+		constexpr ULONG dmfSetTitleBar =        0x00000008UL;
+		// set dark explorer theme
+		// used mainly for dark scrollbar not handled with dmfSetChildren
+		// might change style for other elements
+		// should be handled on first time and in NPPN_DARKMODECHANGED
+		// requires at least Windows 10
+		constexpr ULONG dmfSetExplorerTheme =   0x00000010UL;
+		// standard first time flags for main parent
+		constexpr ULONG dmfMainParent = dmfSetParent | dmfSetChildren | dmfSubclassChildren | dmfSetTitleBar; // 0x000000FUL
+		// at least one of these flag should be used
+		constexpr ULONG dmfRequiredMask = dmfSetParent | dmfSetChildren | dmfSetTitleBar | dmfSetExplorerTheme; // 0x0000001BUL
+		// every flags
+		constexpr ULONG dmfAllMask = dmfSetParent | dmfSetChildren | dmfSubclassChildren | dmfSetTitleBar | dmfSetExplorerTheme; // 0x0000001FUL
+	};

```

Comment:
All these flags are not explicit for people who are not familiar to dark mode.
Could we reduce some flags which won't be used, and make all the flag names simpler and more explicit, with more explanation and examples?

Reply:
Mask flags are used mainly internally.
I am open to better name.

Reply:
@ozone10 
what are the flags which are used by plugins then?  
I suppose that they are dmfAllMask,  dmfRequiredMask  and dmfMainParent 
Could you confirm that?

Reply:
@ozone10 
What's the different between dmfSetParent & dmfSetChildren ?

Reply:
`dmfSetParent` is for edit, listbox, partially combobox and static text controls. They use `WM_CTLCOLOR*` messages, which are processed by their parents.
It might be better to rename it to `dmfSubclassParent`

`dmfSetChildren` currently sets theme for push buttons and some other controls, and allows subclass. I might change how it interacts with `dmfSubclassChildren` to make it easier for plugin's authors.

Reply:
> dmfSetParent is for edit, listbox, partially combobox and static text controls. They use WM_CTLCOLOR* messages, which are processed by their parents.
It might be better to rename it to dmfSubclassParent

So if a dlgHwnd contains a editHwnd, a comboHwnd and a buttonHwnd as children, how can plugin author call `NPPM_DARKMODESUBCLASSANDTHEME` to set dialog in dark mode?

Reply:
they should use 
```C++
::SendMessage(nppData._nppHandle, NPPM_DARKMODESUBCLASSANDTHEME, static_cast<WPARAM>(NppDarkMode::dmfAfterInitParent), reinterpret_cast<LPARAM>(dlgHwnd));
```

Reply:
So why do we need other flags?
What's "Theme" exactly BTW?

Reply:
`dmfAfterInitParent` is combination of `dmfSubclassParent | dmfSubclassChildren | dmfSetTitleBar` flags, but plugin authors might want to handle some controls differently.

Maybe they want only dark title bar and handle other controls themself.

If editHwnd is not child of dlgHwnd but buttonHwnd, then
```c++
::SendMessage(nppData._nppHandle, NPPM_DARKMODESUBCLASSANDTHEME, static_cast<WPARAM>(NppDarkMode::dmfSubclassParent), reinterpret_cast<LPARAM>(buttonHwnd));
```
should be used.

Also `dmfSubclassParent` should be used only once for required control (parent) and `dmfSubclassChildren` only once for main hwnd.


For theme, Notepad++ uses wrapper for https://learn.microsoft.com/en-us/windows/win32/api/uxtheme/nf-uxtheme-setwindowtheme. And controls need to apply new theme each time dark mode has changed, or they will remain in previous mode and have wrong colors.

Same for title bar.

Reply:
> Maybe they want only dark title bar and handle other controls themself.

That doesn't make sense. ppl handle whole things themself, or they use API completely to do all stuffs.
I have no problem to provide more control to users. But easier & simpler the API is, more ppl will use it. That's my point.

 > If editHwnd is not child of dlgHwnd but buttonHwnd, then...
 
 >  Also dmfSubclassParent should be used only once for required control (parent) and dmfSubclassChildren only once for main hwnd.
 
 Sorry but I don't follow you. Could you elaborate it please?
 
 
 > And controls need to apply new theme each time dark mode has changed, or they will remain in previous mode and have wrong colors.
Same for title bar.

Theme & title bar could be included and hidden in the API, or I'm too naive? For me, plugin authors don't care what it is theme - all they want is their dialog can follow Notepad++'s mode (dark or light mode).

Reply:
> That doesn't make sense. ppl handle whole things themself, or they use API completely to do all stuffs.
> I have no problem to provide more control to users. But easier & simpler the API is, more ppl will use it. That's my point.

This why I provide combined `dmfAfterInitParent` and `dmfHandleChangeParent` flags, if they want to have more granular control they could use individual flags.

They might also use some non-standard controls/parent-child relationship which are not handled by combined flags.

> Sorry but I don't follow you. Could you elaborate it please?
Control should be subclassed only once.

Check related issue https://github.com/npp-plugins/plugindemo/issues/4 for edit control and its parent interaction.

> Theme & title bar could be included and hidden in the API, or I'm too naive? For me, plugin authors don't care what it is theme - all they want is their dialog can follow Notepad++'s mode (dark or light mode).

The problem is that control has no idea that dark mode changed.

Reply:
@ozone10 
The following issue is not addressed: 
https://community.notepad-plus-plus.org/assets/uploads/files/1682252209282-notepad_plugin_darkmode.png

is it?

Reply:
Since this window does not have title bar using `dmfSubclassParent | dmfSubclassChildren` flags is enough to apply dark mode.

And because there is only one button we can use `dmfSetThemeDirectly` in `NPPN_DARKMODECHANGED` on it to handle dark mode change.

Reply:
![image](https://github.com/notepad-plus-plus/notepad-plus-plus/assets/90293/cb2f6e51-67cd-48a1-a9d6-67fd2ddcf490)

```cpp
void about()
{
	HWND hSelf = ::CreateDialogParam(_hInst, MAKEINTRESOURCE(IDD_ABOUTBOX), nppData._nppHandle, (DLGPROC)dlgProc, (LPARAM)NULL);
		    
	// Go to center
	RECT rc;
	::GetClientRect(nppData._nppHandle, &rc);
	POINT center;
	int w = rc.right - rc.left;
	int h = rc.bottom - rc.top;
	center.x = rc.left + w/2;
	center.y = rc.top + h/2;
	::ClientToScreen(nppData._nppHandle, &center);

	RECT dlgRect;
	::GetClientRect(hSelf, &dlgRect);
	int x = center.x - (dlgRect.right - dlgRect.left)/2;
	int y = center.y - (dlgRect.bottom - dlgRect.top)/2;

	::SetWindowPos(hSelf, HWND_TOP, x, y, (dlgRect.right - dlgRect.left), (dlgRect.bottom - dlgRect.top), SWP_SHOWWINDOW);

	::SendMessage(nppData._nppHandle, NPPM_DARKMODESUBCLASSANDTHEME, static_cast<WPARAM>(NppDarkMode::dmfSubclassParent | NppDarkMode::dmfSubclassChildren), reinterpret_cast<LPARAM>(hSelf));
}
```

Reply:
@donho 
`::SetWindowPos` should be last. `NPPM_DARKMODESUBCLASSANDTHEME` requires redraw if window is visible 

![image](https://github.com/notepad-plus-plus/notepad-plus-plus/assets/55940305/39635808-ca88-4a10-8db9-d9cbf632e5f0)

Reply:
Thank you @ozone10